### PR TITLE
RXI-428 atestation for door master/regular keys

### PR DESCRIPTION
**Pull request by oleks-rip**
If witness federator id  is equal to master(and is not disabled)/regular  key of the Door account the attestation proceed even if federator id is not in signer list.

### CODE REVIEW

File: src/xbwd/client/ChainListener.h [hunk 1]
```diff
@@ -53,7 +53,8 @@ class ChainListener : public std::enable_shared_from_this<ChainListener>
     mutable std::mutex callbacksMtx_;
 
     using RpcCallback = std::function<void(Json::Value const&)>;
-    std::map<std::uint32_t, RpcCallback> GUARDED_BY(callbacksMtx_) callbacks_;
+    std::unordered_map<std::uint32_t, RpcCallback> GUARDED_BY(callbacksMtx_)
```
Comment (seelabs) [inline src/xbwd/client/ChainListener.h:56]: Need to `#include <unordered_map>` (`map` should have been included before this change, and wasn't)

File: src/xbwd/client/ChainListener.cpp [hunk 2]
```diff
@@ -845,233 +856,359 @@ ChainListener::processMessage(Json::Value const& msg)
...
+    catch (...)
+    {
+        exc_logger("");
+    }
 }
 
-std::optional<event::XChainSignerListSet>
+std::optional<std::unordered_set<ripple::AccountID>>
 processSignerListSetGeneral(
+    beast::Journal const j,
```
Comment (seelabs) [inline src/xbwd/client/ChainListener.cpp:893]: Journal is always the last parameter. Also, in general we don't pass value parameters by const (i.e. `int`, not `int const`).

File: src/xbwd/client/ChainListener.cpp [hunk 3]
```diff
@@ -845,233 +856,359 @@ ChainListener::processMessage(Json::Value const& msg)
...
+            WARN_RET("'result' missed");
 
+        auto const& jres = msg[ripple::jss::result];
         if (!jres.isMember(ripple::jss::account_data))
-        {
-            logBadMsg(j_, "'accoujnt_data' missed", msg, chainName);
-            return;
-        }
-        const auto& jaccData = jres[ripple::jss::account_data];
+            WARN_RET("'accoujnt_data' missed");
```
Comment (seelabs) [inline src/xbwd/client/ChainListener.cpp:949]: typo: `accoujnt`

File: src/xbwd/client/ChainListener.cpp [hunk 4]
```diff
@@ -845,233 +856,359 @@ ChainListener::processMessage(Json::Value const& msg)
...
+        JLOGV(
+            j.warn(),
+            errTopic,
+            ripple::jv(
+                "exception", !exc_err.empty() ? exc_err : "unknown exception"),
+            ripple::jv("msg", msg),
+            ripple::jv("chain_name", chainName));
+    };
+
+    exceptionWrapper(worker, exc_logger);
```
Comment (seelabs) [inline src/xbwd/client/ChainListener.cpp:1018]: I probably prefer removing the `worker` lambda and just explicitly showing the try/catch.

**Review by seelabs — COMMENTED** [review]
Did a pass and left a few comments. Looks good tho, thanks @oleks-rip!

File: src/xbwd/federator/Federator.h [hunk 5]
```diff
@@ -70,10 +70,22 @@ struct Submission
...
-    enum class KeySignerListStatus : int { unknown = -1, absent = 0, present };
+    enum KeySignerListStatus : int { unknown = -1, absent = 0, present };
+
+    KeySignerListStatus status_ = KeySignerListStatus::unknown;
+    bool presentInSignerList_ = false;
+    bool ignoreSignerList_ = false;
+    bool disableMaster_ = false;
+    ripple::AccountID regularDoorID_;
+
+    inline Json::Value
```
[the remark this answers is not in the document]
Comment (oleks-rip) [reply src/xbwd/federator/Federator.h:83]: done

File: src/xbwd/client/ChainListener.cpp [hunk 6]
```diff
@@ -380,11 +381,21 @@ ChainListener::processMessage(Json::Value const& msg)
         fieldMatchesStr(msg, ripple::jss::type, ripple::jss::transaction))
     {
         auto const& txn = msg[ripple::jss::transaction];
-        if (fieldMatchesStr(
-                txn, ripple::jss::TransactionType, ripple::jss::SignerListSet))
-        {
+        auto const txType = [](Json::Value const& val) -> std::string {
```
[the remark this answers is not in the document]
Comment (oleks-rip) [reply src/xbwd/client/ChainListener.cpp:384]: It figure out that it is not so easy add these transactions to the switch(663).  There are many checks making prior that switch and trying pass them lead to more code distortion. I propose to leave it to the time when I'll rewrite it all.

File: src/xbwd/client/ChainListener.cpp (hunk 4 above)
Comment (oleks-rip) [reply src/xbwd/client/ChainListener.cpp]: done c8a6e9b86731

File: src/xbwd/client/ChainListener.cpp (hunk 3 above)
Comment (oleks-rip) [reply src/xbwd/client/ChainListener.cpp]: done c8a6e9b86731

File: src/xbwd/client/ChainListener.cpp (hunk 2 above)
Comment (oleks-rip) [reply src/xbwd/client/ChainListener.cpp]: done c8a6e9b86731

File: src/xbwd/client/ChainListener.cpp [hunk 7]
```diff
@@ -845,233 +856,359 @@ ChainListener::processMessage(Json::Value const& msg)
...
 
+#define WARN_RET(err_msg, ...)                    \
+    {                                             \
+        JLOGV(                                    \
+            j.warn(),                             \
+            errTopic,                             \
+            ripple::jv("reason", err_msg),        \
+            ripple::jv("msg", msg),               \
+            ripple::jv("chain_name", chainName)); \
+        return __VA_ARGS__;                       \
```
[the remark this answers is not in the document]
Comment (oleks-rip) [reply src/xbwd/client/ChainListener.cpp]: done c8a6e9b86731

File: src/xbwd/client/ChainListener.h (hunk 1 above)
Comment (oleks-rip) [reply src/xbwd/client/ChainListener.h:56]: done c8a6e9b86731

File: src/xbwd/federator/Federator.h [hunk 8]
```diff
@@ -249,4 +267,21 @@ make_Federator(
     config::Config const& config,
     beast::Journal j);
 
+//-------- implementation ---------
```
[the remark this answers is not in the document]
Comment (oleks-rip) [reply src/xbwd/federator/Federator.h]: done c8a6e9b86731

File: src/xbwd/client/ChainListener.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (seelabs) [reply src/xbwd/client/ChainListener.cpp:384]: As it is it breaks the structure of the function which is:
1) Find the transaction type
2) Get relevant information that is common to many transaction's event handling (transaction hash, delivered amount, ect; if a field is not present it is set as a null opt).
3) Switch on the transaction type and call the right function.

It sounds like the concern is the code at (2). This should be OK. If the transaction doesn't have one of these fields (no `delivered_amount` for `SetAccount` for example) just set it to a null opt.

The one minor issue is that code assume it's responding to transactions - if that isn't true we might want to refactor things so things that are not transactions are handled elsewhere. 

I would like to make this change before we merge this patch.

**Comment by seelabs** [comment]
merged

**Pull request closed without merging**